toolkit/components/filewatcher/NativeFileWatcherLnx.cpp

```diff
@@ -0,0 +1,1073 @@
+﻿#include "NativeFileWatcherLnx.h"
+
+#include "mozilla/Services.h"
+#include "mozilla/UniquePtr.h"
+#include "nsClassHashtable.h"
+#include "nsDataHashtable.h"
+#include "nsIFile.h"
+#include "nsIObserverService.h"
+#include "nsProxyRelease.h"
+#include "nsTArray.h"
+#include "mozilla/Logging.h"
+#include "mozilla/Scoped.h"
+
+#include <sys/inotify.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <csignal>
+#include <cstdlib>
+#include <queue>
+#include <cstring>
+#include <errno.h>
+
+#include <iostream>
+
+#include <NativeFileWatcherCommons.h>
+
+namespace mozilla {
+
+namespace moz_filewatcher {
+
+// Global signal counter for events reported by inotify. Used in event handler.
+static volatile sig_atomic_t mEventWaiting;
+
+/**
+ * A structure to hold the information about a single inotify watch descriptor.
+ */
+struct WatchedResourceDescriptor
+{
+    // The path on the file system of the watched resource.
+    nsString mPath;
+
+    // The watched descriptor returned from inotify_add_watch().
+    int mWatchedResourceDescriptor;
+
+    WatchedResourceDescriptor(const nsAString& aPath, const int anHandle)
+        : mPath(aPath)
+        , mWatchedResourceDescriptor(anHandle)
+    {
+    }
+};
+
+/**
+ * This runnable is dispatched from the main thread to get the notifications of
+ * the changes in the watched resources by handling and acting on signals from
+ * the kernel.
+ *
+ * This is accomplished with non-blocking reads from an inotify file descriptor
+ * when the signal handler tied to the inotify file descriptor increments an
+ * event counter. Callbacks are then dispatched depending on the type of event
+ * the inotify_event read from the file descriptor describes.
+ */
+class NativeFileWatcherIOTask : public Runnable
+{
+public:
+    explicit NativeFileWatcherIOTask(int inotifyFileDescriptor)
+        : Runnable("NativeFileWatcherIOTask")
+        , mShuttingDown(false)
+        , mInotifyFileDescriptor(inotifyFileDescriptor)
+    {
+    }
+
+    NS_IMETHOD Run() override;
+    nsresult AddPathRunnableMethod(
+            PathRunnablesParametersWrapper* aWrappedParameters);
+    nsresult RemovePathRunnableMethod(
+            PathRunnablesParametersWrapper* aWrappedParameters);
+    nsresult DeactivateRunnableMethod();
+
+private:
+    // Maintain 2 indexes - one by resource path, one by inotify watch descriptor.
```

Comment:
Documentation needed: is the path a directory or a single file? (here and everywhere else where you have a path)

Reply:
@argencore to review.